# Part 1 - Exploratory data analysis

The attached logins.json file contains (simulated) timestamps of user logins in a particular
geographic location. Aggregate these login counts based on 15minute
time intervals, and
visualize and describe the resulting time series of login counts in ways that best characterize the
underlying patterns of the demand. Please report/illustrate important features of the demand,
such as daily cycles. If there are data quality issues, please report them.

In [1]:
# Import python packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
file_name_path = 'data/logins.json'

In [3]:
# Loading json file into pandas data frame
login_df = pd.read_json(file_name_path)

In [4]:
# check first few lines of data
login_df.head()

login_time
0 1970-01-01 20:13:18
1 1970-01-01 20:16:10
2 1970-01-01 20:16:37
3 1970-01-01 20:16:36
4 1970-01-01 20:26:21

In [5]:
login_df.shape

(93142, 1)

In [6]:
# Verify type of column
# https://forum.onefourthlabs.com/t/regarding-datatype-dtype-m8-ns/7189
login_df['login_time'].dtypes

dtype('<M8[ns]')

In [7]:
login_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93142 entries, 0 to 93141
Data columns (total 1 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   login_time  93142 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 727.8 KB


In [8]:
# Describe dataframe
login_df.describe(datetime_is_numeric=True)

login_time
count                          93142
mean   1970-02-26 19:09:39.779648278
min              1970-01-01 20:12:16
25%       1970-02-04 10:37:19.750000
50%       1970-03-01 06:33:05.500000
75%       1970-03-22 04:01:10.750000
max              1970-04-13 18:57:38

In [9]:
# Unique values
login_df.nunique()

login_time    92265
dtype: int64

In [10]:
login_df.value_counts().head()

login_time         
1970-04-12 04:03:15    3
1970-02-12 11:16:53    3
1970-03-14 17:45:52    3
1970-01-06 21:45:52    3
1970-04-04 01:17:40    3
dtype: int64

In [11]:
login_count = login_df.login_time.value_counts().to_frame()

In [12]:
login_count.login_time.value_counts()

1    91397
2      859
3        9
Name: login_time, dtype: int64

### In total 93142 records ranging from 1970 Jan 1st 20:12 to April 13th 18:57, 859 timestamps are doubled and 9 records are trippled and it can be multiple user logins.?!

In [13]:
login_df.sort_values(by="login_time", inplace=True)

In [14]:
# create new column year and month for easy visualization
login_df['year'] = login_df.login_time.dt.year
login_df['month'] = login_df.login_time.dt.month
login_df['day'] = login_df.login_time.dt.day
login_df['day_name'] = login_df.login_time.dt.day_name()
login_df['month_name'] = login_df.login_time.dt.month_name()
login_df['week'] = login_df.login_time.dt.week

<ipython-input-14-5540074a0606>:7: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  login_df['week'] = login_df.login_time.dt.week


In [15]:
login_df.head()

login_time  year  month  day  day_name month_name  week
6 1970-01-01 20:12:16  1970      1    1  Thursday    January     1
0 1970-01-01 20:13:18  1970      1    1  Thursday    January     1
1 1970-01-01 20:16:10  1970      1    1  Thursday    January     1
3 1970-01-01 20:16:36  1970      1    1  Thursday    January     1
2 1970-01-01 20:16:37  1970      1    1  Thursday    January     1

In [16]:
# count Login by month
login_df.groupby(['month', 'month_name'])['day'].count()

month  month_name
1      January       21239
2      February      24673
3      March         33723
4      April         13507
Name: day, dtype: int64

### April having less data compared to previous month

In [18]:
login_by_day=login_df.groupby(['month_name', 'month', 'day'])['day'].count().to_frame(name='login').reset_index()

In [21]:
# count days per month
login_by_day.groupby(['month_name', 'month'])['day'].count()

month_name  month
April       4        13
February    2        28
January     1        31
March       3        31
Name: day, dtype: int64

### Looks like april having only 13 days entry

In [22]:
login_by_day.sample()

month_name  month  day  login
87      March      3   16    837

In [23]:
login_by_day[login_by_day.month_name == 'April']

month_name  month  day  login
0       April      4    1    939
1       April      4    2   1127
2       April      4    3   1504
3       April      4    4   1889
4       April      4    5   1162
5       April      4    6    615
6       April      4    7    590
7       April      4    8    698
8       April      4    9    818
9       April      4   10   1009
10      April      4   11   1352
11      April      4   12   1409
12      April      4   13    395